source/dofs/dof_tools_sparsity.cc

```diff
@@ -409,24 +413,25 @@ namespace DoFTools
             // make sparsity pattern for this cell
             for (unsigned int i = 0; i < dofs_per_face; ++i)
               for (unsigned int j = 0; j < dofs_per_face; ++j)
-                sparsity.add(dof_to_boundary_mapping[dofs_on_this_face[i]],
-                             dof_to_boundary_mapping[dofs_on_this_face[j]]);
+                append_pair(dof_to_boundary_mapping[dofs_on_this_face[i]],
+                            dof_to_boundary_mapping[dofs_on_this_face[j]]);
+
+            sparsity.add_entries(make_array_view(rows), make_array_view(cols));
```

Comment:
This is maybe a design flaw, but there ought to be a more efficient way for this. You're building the tensor product of an index set with itself, and so add a dense sub-matrix to the sparsity pattern. It doesn't make particular sense to build the full `N*N` set of indices here. One should be able to get away with `2*N` or `N`, depending on whether you want to pass two or one index sets.

Separately, though, there is also the question that conceptually it is awkward to pass two arrays of row and column indices when really you want to pass one array of `(row,column)` pairs. I recognize that this happened in a previous step of your patch series, but wanted to mention it before it gets all baked into the library for good.

Reply:
I agree, having a possibility to pass in two arrays to build a tensor product from would be a nice interface to have.

Reply:
The main complexity in these patches was figuring out how to make things work with `AffineConstraints`: making more interface changes right now won't be difficult and I am happy to do it.

I'm ambivalent about whether we should pass a pair of arrays or an array of pairs. I chose "a pair of arrays" to make the base class method resemble the interface for `add()`. All that these patches do is translate `add()` to `add_entries(rows, columns)` (I did it with simple search-and-replace). One possibility would be to upgrade this code to use `add_row_entries()` (i.e., pass `N` indices `N` times). I suspect this code was written before the more-optimized `add_row_entries(row, start, end, is_sorted)` interface was added to the sparsity pattern classes anyway.

If we implemented `add_tensor_product(r1, r2, c1 , c2)` it would still need to, at some point, form ranges of columns to work with `DynamicSparsityPattern::Line` and `BlockSparsityPatternBase` or duplicate their additional indexing logic. Hence I think it makes more sense to upgrade the present code to `add_row_entries()`.

Reply:
I spent about 20 minutes converting some of these to `add_row_entries()` - the result is much nicer.